src/jit/lsra.cpp

```diff
@@ -6462,8 +6655,30 @@ LinearScan::allocateRegisters()
 #endif // FEATURE_SIMD
                 if (currentRefPosition->RequiresRegister())
                 {
-                    assignedRegister = allocateBusyReg(currentInterval, currentRefPosition);
-                    INDEBUG(dumpLsraAllocationEvent(LSRA_EVENT_ALLOC_SPILLED_REG, currentInterval, assignedRegister));
+                    if (allocateReg)
+                    {
+                        // Though Lower/Codegen has indicated that it can generate code even if
+                        // no reg is allocated to this ref position, we will make an attempt
+                        // to get a busy reg if it is allocated to a lesser important ref position.
+                        // If all the refpositions currently occupying registers are more
+                        // important than currentRefPosition, no reg will be allocated.                         
+                        assignedRegister = allocateBusyReg(currentInterval, currentRefPosition, currentRefPosition->IsRegOptional());
+                    }
```

Comment:
As discussed offline, for this PR I am going to rename RequiresRegister() to IsActualRef().  After merging this PR, I will attempt your second suggestion as it would be a pure restructuring change with minor logic addition.